src/components/Contacts/ContactDetails/ContactDetailsTab/Tags/ContactTags.tsx

```diff
@@ -15,43 +24,189 @@ const ContactTagChip = styled(Chip)(({ theme }) => ({
+}));
+
+const ContactTagInput = styled(TextField)(({ theme }) => ({
+  '&& .MuiFormLabel-root.Mui-focused': {
+    color: theme.palette.cruGrayDark.main,
+  },
+  '&& .MuiInput-underline:before ': {
+    borderBottom: 'none',
+  },
+  '&& .MuiInput-underline:after ': {
+    borderBottom: `2px solid ${theme.palette.divider}`,
+  },
+  '&& .MuiInputBase-input': {
+    marginBottom: '13px',
+  },
   margin: theme.spacing(1),
   marginLeft: '0',
 }));
 
 interface ContactTagsProps {
+  accountListId: string;
   contactId: string;
   contactTags: string[];
 }
 
+const tagSchema = yup.object({
+  tag: yup.string().required(),
+});
+
 export const ContactTags: React.FC<ContactTagsProps> = ({
+  accountListId,
   contactId,
   contactTags,
 }) => {
   const { t } = useTranslation();
+  const { enqueueSnackbar } = useSnackbar();
+  const [
+    updateContactTags,
+    { loading: updating },
+  ] = useUpdateContactTagsMutation();
+
   const handleTagDelete = (tag: string) => {
     const index = contactTags.indexOf(tag);
+
     if (index > -1) {
-      contactTags = contactTags.splice(index, 1);
-      // TODO: add mutation to update tag for contact
+      const tagList = [...contactTags];
+      tagList.splice(index, 1);
+      try {
+        updateContactTags({
+          variables: {
+            accountListId,
+            contactId,
+            tagList,
+          },
+          update: (cache, { data: updateData }) => {
+            const query = {
+              query: ContactDetailsTabDocument,
+              variables: {
+                accountListId,
+                contactId,
+              },
+            };
+
+            const dataFromCache = cache.readQuery<ContactDetailsTabQuery>(
+              query,
+            );
+
+            if (dataFromCache) {
+              const data = {
+                contact: {
+                  ...dataFromCache.contact,
+                  tagList: updateData?.updateContact?.contact.tagList,
+                },
+              };
+              cache.writeQuery({ ...query, data });
+              enqueueSnackbar(t('Tag successfully removed'), {
+                variant: 'success',
+              });
+            }
+          },
+        });
+      } catch (error) {
+        enqueueSnackbar(error.message, { variant: 'error' });
+        throw error;
+      }
     }
-    console.info('You clicked the Delete tag for tag' + tag + '' + contactId);
   };
-  const handleAddTag = () => {
-    // TODO: add mutation to update tag for contact
-    console.info('You clicked add contact Tag to contact' + contactId);
+
+  const onSubmit = async (
+    { tag }: { tag: string },
+    { resetForm }: FormikHelpers<{ tag: string }>,
+  ): Promise<void> => {
+    try {
+      resetForm();
+      updateContactTags({
+        variables: {
+          accountListId,
+          contactId,
+          tagList: [...contactTags, tag],
+        },
+        optimisticResponse: {
+          updateContact: {
+            contact: {
```

Comment:
This was throwing runtime errors about `Invariant Violation: Missing field '__typename' `
```suggestion
            __typename: 'ContactUpdateMutationPayload',
            contact: {
              __typename: 'Contact',
```
solves it. Idk if there's a better way to do that.

Reply:
Do you think using an optimistic response is worth it? I've seen it in a few other places but wasn't sure if we wanted to continue to use it. Those other spots don't have the ```_typename``` as well so I'm curious if they are throwing a similar errror.

Reply:
I think optimistic responses are a great feature. They are a little more work to write in maintain but hopefully it's not much. Are they strictly typed?